File: services/appointment-service/models/appointment.go
package models

import "time"

type Status string

const (
	StatusScheduled  Status = "scheduled"
	StatusCheckedIn  Status = "checked_in"
	StatusInProgress Status = "in_progress"
	StatusCompleted  Status = "completed"
	StatusCancelled  Status = "cancelled"
	StatusNoShow     Status = "no_show"
)

type Appointment struct {
	ID            string     `json:"id"`
	PatientID     string     `json:"patient_id"`
	DoctorID      *string    `json:"doctor_id"`      // null for session-based bookings
	StartTime     *time.Time `json:"start_time"`     // null for session-based bookings
	Session       *string    `json:"session"`        // "morning" | "afternoon" | null
	EstimatedTime *time.Time `json:"estimated_time"` // set by ETA service
	QueuePosition *int       `json:"queue_position"` // set by queue coordinator
	Notes         *string    `json:"notes"`
	Status        Status     `json:"status"`
	CreatedAt     time.Time  `json:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at"`
}

type CreateAppointmentRequest struct {
	PatientID string     `json:"patient_id" binding:"required"`
	DoctorID  *string    `json:"doctor_id"`
	StartTime *time.Time `json:"start_time"` // required for specific doctor bookings
	Session   *string    `json:"session"`    // required for generic bookings: "morning" | "afternoon"
	Notes     *string    `json:"notes"`
}

type Doctor struct {
	ID             string `json:"id"`
	Name           string `json:"name"`
	Specialization string `json:"specialization"`
	SlotCapacity   int    `json:"slot_capacity"`
}

type UpdateStatusRequest struct {
	Status Status `json:"status" binding:"required"`
}

File: composite/appointment/src/models/appointment.py
from pydantic import BaseModel, model_validator
from datetime import datetime
from typing import Literal


# ─── Incoming from frontend ──────────────────────────────────

class CreateAppointmentRequest(BaseModel):
    patient_id: str
    doctor_id: str | None = None
    start_time: datetime | None = None                       # required for specific doctor bookings
    session: Literal["morning", "afternoon"] | None = None  # required for generic bookings
    notes: str | None = None

    @model_validator(mode="after")
    def validate_booking_type(self):
        if self.session and self.start_time:
            raise ValueError("provide either session or start_time+doctor_id, not both")
        if not self.session and not self.start_time:
            raise ValueError("provide either session (morning/afternoon) or start_time+doctor_id")
        if self.start_time and not self.doctor_id:
            raise ValueError("doctor_id is required when start_time is provided")
        return self


# ─── Sent to atomic appointment-service ──────────────────────

class AppointmentServiceRequest(BaseModel):
    patient_id: str
    doctor_id: str | None = None
    start_time: datetime | None = None
    session: str | None = None
    notes: str | None = None


# ─── Sent to downstream services (via RabbitMQ) ──────────────

class AppointmentBookedEvent(BaseModel):
    appointment_id: str
    patient_id: str
    doctor_id: str | None = None
    start_time: datetime | None = None
    session: str | None = None


# ─── Returned to frontend ────────────────────────────────────

class AppointmentResponse(BaseModel):
    id: str
    patient_id: str
    doctor_id: str | None = None
    start_time: datetime | None = None
    session: str | None = None
    estimated_time: datetime | None = None  # set later by ETA service
    queue_position: int | None = None       # set later by queue coordinator
    notes: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime
